swig/python/codegen/codegen.py

```diff
@@ -0,0 +1,1172 @@
+from __future__ import print_function, unicode_literals
+import six
+import sys
+import ast
+import os
+import tokenize
+import warnings
+from six import StringIO
+
+
+def interleave(inter, f, seq):
+    """Call f on each item in seq, calling inter() in between.
+    """
+    seq = iter(seq)
+    try:
+        f(next(seq))
+    except StopIteration:
+        pass
+    else:
+        for x in seq:
+            inter()
+            f(x)
+
+class CodeGenException(Exception):
+    """ Generic exception for errors raised in code generation """
+    pass
+
+class CodeGenerator:
+    """Methods in this class recursively traverse an AST and
+    output source code for the abstract syntax; original formatting
+    is disregarded. """
+
+    # represents built in functions
+    pythonbuiltins = ["abs", "float", "int"]
+    
+    # basic types
+    basic_arg_types = ['float', 'int']
+    
+    # supported math constansts    
+    mathconsts = {"pi": "M_PI",
```

Comment:
Why is the other python `math` constatant `math.tau` not supported? ([docs](https://docs.python.org/3/library/math.html#math.tau)). 

It's available as python 3.6, which is currently our minimum but moving to >= 3.7 due to 3.6 being EOL, so it should be safe to include this.

Reply:
Because math.h does not have a definition for `TAU`

Reply:
Fair, might be worth adding to the docs then that `math.tau` is not supprted, but can be computed as `2 * math.pi`

Reply:
Will do.